### ISSUE TITLE
Pharo broken on 17.03

### ISSUE THREAD

**Issue by lukego** [NixOS/nixpkgs#24541]
## Issue description

I am not able to run Pharo on NixOS with the 17.03 channel. Running `pharo-launcher` I get a segmentation fault with a long backtrace ([gist](https://gist.github.com/lukego/29cffe9b514b569049f7691ad7a602b7)). (Stack overflow?)

I also tried running the latest Pharo 5 VM from pharo.org (`Pharo-50771.image`) using the `pharo-spur-vm-x` from nixpkgs and this failed with a different error ([gist](https://gist.github.com/5d86fc3c9b6c8f25a807fab309b874fb)).

Any ideas @DamienCassou?

## Technical details

* System: NixOS 16.03.1349.dda40aa8d1 (Emu)
* Nix version: nix-env (Nix) 1.11.4
* Nixpkgs version: 17.03.873.0c041520c3.

**Comment by lukego** [comment]
Just more detail: I wanted to see whether the X server could be a factor but I have tested running both under TigerVNC (`vncserver`) and also with SSH forwarding X11 to an OSX Mavericks host and the problem is the same.

I also tried running the stock Pharo 5.0 image in a headless spur VM and that fails like this:

```
[luke@interlaken:~]$ pharo-spur-vm-nox Pharo-50771.image 
Array>>exception
Array(Object)>>doesNotUnderstand: #exception
Context>>runUntilErrorOrReturnFrom:
[ tmp1 runUntilErrorOrReturnFrom: arg1 ] in Process>>complete: in Block: [ tmp1 runUntilErrorOrReturnFrom: arg1 ]
BlockClosure>>ensure:
Process>>evaluate:onBehalfOf:
Process>>complete:
Process>>return:value:
[ self return: tmp1 value: tmp1 receiver ] in Process>>popTo: in Block: [ self return: tmp1 value: tmp1 receiver ]
BlockClosure>>ensure:
Process>>evaluate:onBehalfOf:
Process>>popTo:
Process>>terminate
WeakArray class>>restartFinalizationProcess
WeakArray class>>startUp:
ClassSessionHandler>>startup:
[ :arg2 | arg2 startup: arg1 ] in WorkingSession>>runStartup: in Block: [ :arg2 | arg2 startup: arg1 ]
[ arg2 value: arg3 ] in [ :arg3 | 
[ arg2 value: arg3 ]
	on: Exception
	do: [ :arg4 | self errorHandler handleError: arg4 ] ] in WorkingSession>>runList:do: in Block: [ arg2 value: arg3 ]
BlockClosure>>on:do:
[ :arg3 | 
[ arg2 value: arg3 ]
	on: Exception
	do: [ :arg4 | self errorHandler handleError: arg4 ] ] in WorkingSession>>runList:do: in Block: [ :arg3 | ...
Array(SequenceableCollection)>>do:
WorkingSession>>runList:do:
WorkingSession>>runStartup:
WorkingSession>>start:
SessionManager>>snapshot:andQuit:
SmalltalkImage>>snapshot:andQuit:
ImageCleanerCommandLineHandler>>activate
ImageCleanerCommandLineHandler class(CommandLineHandler class)>>activateWith:
[ arg1 activateWith: commandLine ] in PharoCommandLineHandler(BasicCommandLineHandler)>>activateSubCommand: in Block: [ arg1 activateWith: commandLine ]
BlockClosure>>on:do:
PharoCommandLineHandler(BasicCommandLineHandler)>>activateSubCommand:
receiver of "last" is nil
UndefinedObject(Object)>>doesNotUnderstand: #last
[ :arg1 | (tmp4 := self lastMatchValueSequence: {arg1}) notNil ] in FreeTypeNameParser>>extractSlant in Block: [ :arg1 | (tmp4 := self lastMatchValueSequence: {a...etc...
[ :arg4 | 
(arg1 value: arg4)
	ifTrue: [ ^ arg2 cull: arg4 ] ] in Array(Collection)>>detect:ifFound:ifNone: in Block: [ :arg4 | ...
Array(SequenceableCollection)>>do:
Array(Collection)>>detect:ifFound:ifNone:
Array(Collection)>>detect:ifNone:
FreeTypeNameParser>>extractSlant
FreeTypeNameParser>>parse
FreeTypeEmbeddedFileInfo(FreeTypeFileInfoAbstract)>>extractAttributesFromNames
EmbeddedFreeTypeFontInstaller>>addFromFileContents:baseName:
SourceCodeProRegular class(EmbeddedFreeTypeFontFontDescription class)>>installFontsIn:
[ :arg2 | arg2 installFontsIn: arg1 ] in EmbeddedFreeTypeFontFontDescription class>>installAllFontsIn: in Block: [ :arg2 | arg2 installFontsIn: arg1 ]
OrderedCollection>>do:
EmbeddedFreeTypeFontFontDescription class>>installAllFontsIn:
EmbeddedFreeTypeFontInstaller>>installAllFontsIn:
[ :arg1 | arg1 installAllFontsIn: self ] in FreeTypeFontProvider>>updateEmbeddedFreeTypeFonts in Block: [ :arg1 | arg1 installAllFontsIn: self ]
OrderedCollection>>do:
FreeTypeFontProvider>>updateEmbeddedFreeTypeFonts
FreeTypeSettings class>>updateFreeType
FreeTypeSettings class>>startUp:
ClassSessionHandler>>startup:
...
```

**Comment by DamienCassou** [comment]
I'm very sorry about all your troubles. I recently tried several ways myself and miserably failed too (even on Fedora). As I'm not using Nix nor Pharo anymore, I won't be able to help you. May I suggest asking the pharo-dev mailing list? Good luck.
-- 
Damien Cassou
http://damiencassou.seasidehosting.st

"Success is the ability to go from one failure to another without
losing enthusiasm." --Winston Churchill

**Comment by lukego** [comment]
No worries! Thanks for confirming that there is a problem and I am not just goofing up :). I'll try to track it down and propose a fix.

**Comment by rydnr** [comment]
I have the same issue, and filled-in an issue in Pharo issue tracker. However, it's difficult to set up an environment to reproduce it, since it requires a NixOS installation.
My plan is to prepare such an environment and help fixing the issue.
Meanwhile, I launch pharo in a Docker environment, and sharing the X windows from the host.

**Comment by lukego** [comment]
I can see that `pharo-launcher` is working in nixpkgs `15.03` but failing with `(Segmentation fault)` message on `master`. So I am running a `git bisect` to identify which commit introduced the segmentation fault. Could be a specific new version of some library, for example?

```shell
cat > pharo-nix-bisect.sh <<EOF
#!/usr/bin/env bash
nix-env -j 10 -f . -iA pkgs.pharo-launcher || exit 125
timeout --preserve-status 20 pharo-launcher | grep '(Segmentation fault)'
status=$?
if [ "$status" == 0 ]; then
    echo "SEGFAULT"
    exit 1
else
    echo "OK"
    exit 0
fi
EOF
chmod +x pharo-nix-bisect.sh

git bisect start master 15.09
git bisect run ./pharo-nix-bisect.sh
```

Seems like git-bisect+nixpkgs is a match made in heaven, since every single change to every single dependency can be automatically searched for the breakage in logarithmic time, but in practice it is running very slowly with a lot of local compilation.

Perhaps Hydra binary cache has already flushed the 15.09 era and I should have manually looked for another starting point? Oh well I will leave it to run in the background and not worry too much about the constant factors for now.

**Comment by lukego** [comment]
@rydnr The fogbugz link does not seem to lead to an issue for me...?

**Comment by lukego** [comment]
The `git bisect` has completed: [git-bisect-log.txt](https://gist.github.com/lukego/2d43d3a1b2e273f03849cae19cb90f4a). The trouble is that many versions were "Skipped" because they couldn't be built e.g. because the source URL referenced in the older nixpkgs version is not valid. (Would be nice if Hydra would cache sources for 5+ years and/or migrate them onto ipfs... has this been discussed?)

Overall the theme of the suspicious commits seems to be "hardening." I will check if there is some obvious solution based on the content of those commits. Otherwise perhaps have to find a way to make them build somehow e.g. yuckery with applying patches to make the sources findable.

**Comment by lukego** [comment]
Got it. The problem is `pic` hardening and I have pushed a fix at #24595.

**Comment by lukego** [comment]
@rydnr You might want to mention to upstream that the problem seems to be related to compiling the VM with the GCC option `-fPIC`. I am not sure if this is supposed to be supported with Pharo or not.
